File: 数据结构和算法/python/quickSort.py
# 快速排序的实现

def quickSort(array, q, r):
    if (q >= r):
        return 
    p = partition(array, q, r)
    quickSort(array, q, p - 1)
    quickSort(array, p + 1, r)

def partition(array, q, r):
    x, y = q, q
    while x < r:
        if (array[x] <= array[r]):
            array[x], array[y] = array[y], array[x]
            y += 1
        x += 1
    array[x], array[y] = array[y], array[x]
    return y


if __name__ == "__main__":
    array = [1,5,3,6,-1,22,6, -2, 11]
    quickSort(array, 0, len(array) - 1)
    print (array)

File: 数据结构和算法/java/QuickSort.java
public class QuickSort{

    public void sort(int[] array, int startIndex, int endIndex){
        if (startIndex >= endIndex){
            return;
        }
        int p = partition(array, startIndex, endIndex);
        sort(array, startIndex, p - 1);
        sort(array, p + 1, endIndex);
    }

    public int partition(int[] array, int startIndex, int endIndex){
        int x = startIndex;
        int y = startIndex;
        for (; x <= endIndex; x++){
            if (array[x] <= array[endIndex]){
                int tmp = array[x];
                array[x] = array[y];
                array[y] = tmp;
                y++;
            }
        }
        return --y;
    }


    public static void main(String[] args) {
        QuickSort quickSort = new QuickSort();
        int[] array = new int[]{1, 2, -1, 8, -7 , 5, 4, 9};
        quickSort.sort(array, 0, array.length - 1);
        for (int s : array){
            System.out.println(s);
        }
    }


}